docs/proposals/resource-quota.md

```diff
@@ -0,0 +1,404 @@
+  // If empty, the quota matches every resource.
+  FieldSelector *FieldSelector `json:"fieldSelector,omitempty"`
+}
+```
+
+The selectors are immutable post-creation.
+
+API definition of `ReplenishmentPolicy` to support replenishment
+via a rate are deferred pending further prototyping to understand
+what additional fields may or may not be needed to track when
+quota was last replenished.
+
+### Pod
+
+The administrator is able to filter terminating vs non-terminating workloads
+by defining a field selector of `spec.restartPolicy`.
+
+The administrator is not able to filter pods by quality of service.  To
+facilitate this ability, the `Pod` should expose a new field on it's `PodSpec`
+that corresponds to its `QualityOfService` that is set by the server at
+creation and used by `ResourceQuota`.  The administrator would then
+be able to filter quota by tracking `spec.qualityOfService` to quota
+`BestEffort` pods differently than `Burstable` and `Guaranteed`.
+
+## Implementation Phases and Anticipated Issues
+
+Phase 1:
+
+* Support ability for cluster-admin to quota based on workload duration
+ * Add `spec.restartAlways` as a field selector for `Pod`
+ * Ensure API server support for set-based and equality based matching of field selector
+ * Add `FieldSelector` to `v1` API
+ * Add `FieldSelector` to `ResourceQuotaSpec` to enable tracking on restart policy
+ * Add logic to admission control and quota controller to enable support
+
+* Improve latency of quota controller observing freed resources for non-pod types
+by watching for more resource types
+
+* Identify and solve issues when running large numbers of namespaces and quotas
+ * Target 20k namespaces with 60k quota documents as baseline scenario
+ * Minimize contention on locking of quota documents at admission
+
+Phase 2:
+
+* Support ability for cluster-admin to quota based on quality of service
+ * Add new field to pod that denotes quality of service class
+ * Ensure field is populated as part of `ResourceQuota` admission plug-in based on
+ enumerated resource requirements during create
+ * For operators that do not run `ResourceQuota`, ensure field is calculated in
+ pod strategy
+
+Phase 3:
+
+* Support ability to define replenishment policy to control rate of creation
+  * Add `ReplenishmentPolicy` to `ResourceQuota`
+
+* Measure performance impact of various time-based replenishment policy intervals
+with large numbers of `ResourceQuota` documents (~60k) in a cluster
+  * Target `Daily` replenishment policy as first option
+
+* Minimize controllers that operate on other users behalf using up a user's
+creation rate.  For example, a replication controller that keeps trying to
+scale up a resource may deplete the user's daily creation rate.  Controller's
+should be sensitive to quota by not making requests that would just be rejected
+by quota.  For example, if a quota allowed 10 pods in total, but the user set
```

Comment:
@derekwaynecarr I am very interested in your new mechanism as well as the systems's general behavior when more work is requested than allowed/available.  I think these situations and use cases take us down a path where we will start prioritizing workloads.  With this proposal an administrator could create a fairly complex quota policy that would be difficult for a user of the system to fully understand the impact of various fields and labels on their ability to create new pods.  I would expect users to have a feel for what they would want to run first when running up against a quota but I would expect it would be hard for the current system to make this determination automatically for them.